### PR TITLE
Add help in case ref and mut are switched

### LINKED ISSUE #43286: Create a helpful error message when typing "mut ref" instead of "ref mut"
I've gotten the order of mut and ref mixed up a few times like this:

let mut ref y = &x;

This generates a "error: expected identifier, found keyword `ref`" error which isn't too helpful for people used to languages that allows the order of keywords to be mixed e.g. static and const in C/C++.

The compiler should suggest "did you mean 'ref mut'?" for the above.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #43286.

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -640,8 +640,14 @@ impl<'a> Parser<'a> {
         match self.token {
             token::Ident(i) => {
                 if self.token.is_reserved_ident() {
-                    self.span_err(self.span, &format!("expected identifier, found {}",
-                                                      self.this_token_descr()));
+                    let desc = self.this_token_descr();
+                    if desc == "keyword `ref`" {
```
Comment (arielb1) [inline src/libsyntax/parse/parser.rs]: I think the correct way of doing this is `self.token.is_keyword(keyword::Ref)`

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -640,8 +640,14 @@ impl<'a> Parser<'a> {
...
+                        self.span_err_help(self.span,
```
Comment (oli-obk) [inline src/libsyntax/parse/parser.rs]: I think this is a prime candidate for `span_suggestion`

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/libsyntax/parse/parser.rs]: Good point!

File: src/libsyntax/parse/parser.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/libsyntax/parse/parser.rs]: I wondered for a while between the both of them. But if you prefer suggestion, I'll totally update to it.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by petrochenkov** [comment]
`fn parse_ident` is not a good place for this diagnostics, it will fire in many totally unrelated situations.
Also, no recovery is happening.

I believe the correct fix should look like this:
```diff
--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -3523,8 +3523,18 @@ impl<'a> Parser<'a> {
             }
             // At this point, token != _, &, &&, (, [
             _ => if self.eat_keyword(keywords::Mut) {
-                // Parse mut ident @ pat
-                pat = self.parse_pat_ident(BindingMode::ByValue(Mutability::Mutable))?;
+                // Parse mut ident @ pat / mut ref ident @ pat
+                let mutref_span = self.prev_span.to(self.span);
+                let binding_mode = if self.eat_keyword(keywords::Ref) {
+                    self.diagnostic()
+                        .struct_span_err(mutref_span, "the order of `mut` and `ref` is incorrect")
+                        .span_suggestion(mutref_span, "try switching the order", "ref mut".into())
+                        .emit();
+                    BindingMode::ByRef(Mutability::Mutable)
+                } else {
+                    BindingMode::ByValue(Mutability::Mutable)
+                };
+                pat = self.parse_pat_ident(binding_mode)?;
             } else if self.eat_keyword(keywords::Ref) {
                 // Parse ref ident @ pat / ref mut ident @ pat
                 let mutbl = self.parse_mutability();
```

**Comment by GuillaumeGomez** [comment]
@petrochenkov: Well, you did it. I can take your patch and apply but you sure you don't want to open a PR yourself? I'd feel bad to take it from you. :-/

**Comment by petrochenkov** [comment]
@GuillaumeGomez 
Okay, submitted https://github.com/rust-lang/rust/pull/43489

**Pull request closed without merging**